autovalet_parking/src/autovalet_parking/Parker.py

```diff
@@ -0,0 +1,251 @@
+
+        corrected_quat = Quaternion()
+        corrected_quat.x = x
+        corrected_quat.y = y
+        corrected_quat.z = z
+        corrected_quat.w = w
+
+        return corrected_quat
+
+    def distToGoal(self, goal):
+        base_link_tf = self.tf_buffer.lookup_transform(self.goal_frame_id, # target_frame_id
+                                self.husky_frame_id, #source frame
+                                rospy.Time(0), #get the tf at first available time
+                                rospy.Duration(1.0)) #timeout after 1
+        base_link_x = base_link_tf.transform.translation.x
+        base_link_y = base_link_tf.transform.translation.y
+        goal_x = goal.pose.position.x
+        goal_y = goal.pose.position.y
+        dist = np.linalg.norm(np.array([base_link_x - goal_x, base_link_y - goal_y]))
+
+        return dist
+
+    def tagPoseRANSAC(self):
+        '''
+        distance between quaternions is calculated based on the formula here:
+        https://math.stackexchange.com/questions/90081/quaternion-distance
+        '''
+        quatArray = []
+        for i in range(len(self.tfArray)):
+            quat = self.tfArray[i].transform.rotation
+            quatArray.append(np.array([quat.x, quat.y, quat.z, quat.w]))
+        votes = []
+        for i in range(len(quatArray)):
+            vote = 0
+            for j in range(len(quatArray)):
+                diff = np.abs(1 - np.dot(quatArray[i], quatArray[j]) ** 2)
+                vote = vote + 1 if diff < 0.005 else vote
+            votes.append(vote)
+        # pick the pose with the most inliers
+        tf = self.tfArray[votes.index(max(votes))]
+
+        return tf
+    
+    def calculate_error(self):
+        
+        # Get the ground truth and actual poses
+        gt = rospy.wait_for_message(self.gt_topic, LinkStates)
+        self.target_pose = self.get_target_pose(gt)
+        self.reached_pose = self.get_reached_pose(gt)
+
+        # Error in x and y in cms
+        trans_err = (self.target_pose[:-1] - self.reached_pose[:-1])*100
+        # Error in orientation (yaw) in degrees
+        # make sure the orientation is the same in hardware so that the "270" works as well
+        yaw_err   = np.abs((self.target_pose[2]*180/3.14 - 270) - self.reached_pose[2]*180/3.14) #comparing the aruco tag (rotated by 270deg) with husky's yaw
+        if yaw_err > 300:
+            yaw_err -= 360
+        print "Target pose (x,y,theta): ", self.target_pose[:-1]*100, self.target_pose[2]*180/3.14 - 270
+        print "Reached pose (x,y,theta): ", self.reached_pose[:-1]*100, self.reached_pose[2]*180/3.14
+        print "Translation error (cms): ", trans_err
+        print "Orientation error (deg): ", yaw_err
+        print "Net error (norm of the error)", np.linalg.norm(trans_err)
+    
+    def get_target_pose(self, gt):
+        
+        self.aruco_tag_gt    = gt.pose[-18] # name 'aruco_visual_marker_7::marker' <to-do> find the name! Don't HC
```

Comment:
did we wanna do the finding of the name here?

Reply:
done

Reply:
done